### ISSUE TITLE
Problem finding boost

### ISSUE THREAD

**Issue by VictorEijkhout** [AdaptiveCpp/AdaptiveCpp#1112]
Immediate symptom:
```
-- Could NOT find Boost: missing: fiber (found /work2/00434/eijkhout/boost/installation-boost-1.81.0-clx-gcc9/lib/cmake/Boost-1.81.0/BoostConfig.cmake (found version "1.81.0"))
CMake Error at CMakeLists.txt:38 (message):
  Please provide the path to the root directory of the boost installation
  using the -DBOOST_ROOT option
```
but I'm supplying that option:
```
cmake -D CMAKE_INSTALL_PREFIX=/work2/00434/eijkhout/opensycl/installation-opensycl-0.9.4-clx-gcc9                    
 -D BOOST_ROOT=/work2/00434/eijkhout/boost/installation-boost-1.81.0-clx-gcc9/include 
-D Boost_NO_SYSTEM_PATHS=On 
-D WITH_CUDA_BACKEND=OFF 
-DWITH_ROCM_BACKEND=OFF 
-DWITH_LEVEL_ZERO_BACKEND=OFF 
-D WITH_ACCELERATED_CPU=OFF                      
-D CMAKE_VERBOSE_MAKEFILE=ON                     
-D BUILD_SHARED_LIBS=TRUE                     
/work2/00434/eijkhout/opensycl/opensycl-0.9.4
```

What's in my boost root?
```
[staff opensycl:1704] ls /work2/00434/eijkhout/boost/installation-boost-1.81.0-clx-gcc9/include
boost/
[staff opensycl:1705] ls /work2/00434/eijkhout/boost/installation-boost-1.81.0-clx-gcc9/include/fiber
ls: cannot access /work2/00434/eijkhout/boost/installation-boost-1.81.0-clx-gcc9/include/fiber: No such file or directory
[staff opensycl:1706] ls /work2/00434/eijkhout/boost/installation-boost-1.81.0-clx-gcc9/include/boost/fiber
algo/                 channel_op_status.hpp   detail/              fss.hpp     mutex.hpp       pooled_fixedsize_stack.hpp     recursive_timed_mutex.hpp  type.hpp
all.hpp               condition_variable.hpp  exceptions.hpp       future/     numa/           properties.hpp                 scheduler.hpp              unbuffered_channel.hpp
barrier.hpp           context.hpp             fiber.hpp            future.hpp  operations.hpp  protected_fixedsize_stack.hpp  segmented_stack.hpp        waker.hpp
buffered_channel.hpp  cuda/                   fixedsize_stack.hpp  hip/        policy.hpp      recursive_mutex.hpp            timed_mutex.hpp
```

What's missing?

**Comment by illuhad** [comment]
Hi, fiber and context are not header-only libraries, so please look for a `libboost_fiber.a/so` and `libboost_context.a/so`. Probably in the `lib` subdirectory of the boost root.

**Comment by fodinabor** [comment]
You might want to remove the `include` from the boost root and just use `/work2/00434/eijkhout/boost/installation-boost-1.81.0-clx-gcc9`

**Comment by VictorEijkhout** [comment]
> Hi, fiber and context are not header-only libraries, so please look for a `libboost_fiber.a/so` and `libboost_context.a/so`. Probably in the `lib` subdirectory of the boost root.

```
[staff opensycl:1732] ls /work2/00434/eijkhout/boost/installation-boost-1.81.0-clx-gcc9/lib/lib*fiber*
/work2/00434/eijkhout/boost/installation-boost-1.81.0-clx-gcc9/lib/libboost_fiber.a
```

I can't figure out how to get boost to build an `.so`, but at least there is a static lib.

**Comment by VictorEijkhout** [comment]
> You might want to remove the `include` from the boost root and just use `/work2/00434/eijkhout/boost/installation-boost-1.81.0-clx-gcc9`

Tried, exactly same error.

**Comment by fodinabor** [comment]
Following https://www.boost.org/build/doc/html/bbv2/tutorial/linkage.html just add `link=shared` to the `b2` invocation?

**Comment by blinkfrog** [comment]
// TL;DR: Rebuild boost with "--with-fiber --with-context --with-filesystem --with-atomic --with-test" options.

I confirm the issue. I use Windows. Tried 0.9.4 hipSYCL release, as well as the latest version from github, and various boost versions from 1.75 to 1.83. Fiber component is built, and the .lib file is present.

After setting Boost_DEBUG to ON, I've noticed, that FindBoost correctly finds fiber, but it also writes this just before saying that it didn't find fiber:

"-- Adding boost_fiber dependencies: atomic;context;filesystem;headers"

Although atomic and filesystem were built and I could see corresponding libraries, adding "--with-filesystem --with-atomic" to b2 command (which already had "--with-fiber --with-context --with-test") helped, and FindBoost was able to find fiber component.

To developers: may be add `filesystem` and `atomic` to the list of components searched by `FindBoost` in CMakeLists.txt? At least error message would be more explaining then.

**Comment by VictorEijkhout** [comment]
> // TL;DR: Rebuild boost with "--with-fiber --with-context --with-filesystem --with-atomic --with-test" options.

Do you mean as options for `bootstrap.sh`? It doesn't recognize those options.

**Comment by blinkfrog** [comment]
> > // TL;DR: Rebuild boost with "--with-fiber --with-context --with-filesystem --with-atomic --with-test" options.
> 
> Do you mean as options for `bootstrap.sh`? It doesn't recognize those options.

No, I meant options for `b2` which is created after launching `bootstrap`.

**Comment by Kevin-Delnoije** [comment]
encountered the same problem on ubuntu since I had installed libboost-dev instead of libboost-all-dev.
use 
```bash
sudo apt install libboost-all-dev
```

**Comment by VictorEijkhout** [comment]
> > > // TL;DR: Rebuild boost with "--with-fiber --with-context --with-filesystem --with-atomic --with-test" options.
> No, I meant options for `b2` which is created after launching `bootstrap`.

Still the same error.

**Comment by blinkfrog** [comment]
> Still the same error.

I am not sure actually how if this solution is applicable to your problem, as `b2` is Windows-related, no? Anyway, this is how I build boost now so AdaptiveCpp succesfully builds after. Change the paths according to your preferences.

`b2 toolset=clang-win address-model=64 variant=release,debug --build-type=complete stage --with-fiber --with-context --with-test --with-atomic --with-filesystem --with-headers install --prefix=c:/libraries/boost_1_83_clang17`

**Comment by VictorEijkhout** [comment]
Some of you options such as `build-type=complete` don't work on Linux, so I had to tinker some.

And it's still not finding boost fiber.

**Comment by blinkfrog** [comment]
> Some of you options such as `build-type=complete` don't work on Linux, so I had to tinker some.
> And it's still not finding boost fiber.

I suggest to try to configure AdaptiveCpp with `-DBoost_DEBUG=ON` added to cmake command. This helped me to solve a similar problem. With this command, you get a verbose log: what boost components are found and whether they can be used. In my case, when I had the similar error message that  `fiber` (and I think also `context`) component isn't found, it was shown that it actually found these components, but they also require `atomic`, `filesystem` and `headers` components as dependencies, and, as these are absent, it can't use found `fiber` (and `context`) components, so it reported that they aren't found.

**Comment by frobnitzem** [comment]
I also experienced this, and diagnosed by changing line 95 of AdaptiveCpp/CMakeLists.txt to:

```patch
- find_package(Boost COMPONENTS context fiber ) 
+ find_package(Boost REQUIRED context fiber )
```

This is boost's fault for not correctly documenting dependencies.  I ended up installing boost using the following:
```bash
curl https://archives.boost.io/release/1.89.0/source/boost_1_89_0.tar.bz2 | tar xjf -
module load gcc/12.2.0
cd boost_1_89_0
./bootstrap.sh --with-libraries=fiber,context,atomic,filesystem,test --prefix=$prefix/boost
./b2 install --prefix=$prefix/boost
```
